"dir cd .." does not work

### Subject of the issue
This is the code currently being used to remove one directory level from the current directory
`newDirectory = JTerm.currentDirectory.substring(0, JTerm.currentDirectory.length() - 2);` 
(Dir.java, approximately line 163)
It makes no sense to do it this way. It will only work if the current directory name is 2 characters long

### Your environment
- Windows 10
- Java 8
- dev branch

### Steps to reproduce
type dir cd ..

### Expected behaviour
Path should be updated from
`F:\User\Desktop\Programming\JTerm`
to
`F:\User\Desktop\Programming`

### Actual behaviour
Path is actually updated from
`F:\User\Desktop\Programming\JTerm`
to
`F:\User\Desktop\Programming\JTe`

I changed the `cd ..` command to split the directory path by slashes (depending on the OS) and built the new directory that way.

```
} else if (newDirectory.equals("..")) {
            String directories[];
            if (JTerm.IS_WIN) {
                directories = JTerm.currentDirectory.split("\\\\");
            } else {
                directories = JTerm.currentDirectory.split("/");
            }

            // If there is only one item, then this is the top level directory
            if (directories.length == 1) {
                return;
            }

            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < directories.length - 1; i++) {
                sb.append(directories[i]);
            }

            newDirectory = sb.toString();
```
I'm not aware of any other way to specify a path so I think it should work (?).
I can't seem to run the project though, due to an IndexOutOfBounds Expcetion in JTerm.java.
```
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:123)
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.remove(ArrayList.java:492)
	at jterm.JTerm.initCommands(JTerm.java:142)
	at jterm.JTerm.<clinit>(JTerm.java:68)
	... 3 more
```

--- a/changelog.txt
+++ b/changelog.txt
@@ -1,3 +1,8 @@
+(#123)[3/1/2018-09:32 0.7.0 wuhugh]
+    Closes #72
+        - cd should work properly on Windows now
+    OS type is no longer automatically set to UNIX from Terminal.java
+
 (#123)[11/27/2017-18:12 0.7.0 Sergix]
 	Merged #104.
 		- Thanks to @DataSecs!

--- a/src/main/java/jterm/command/Dir.java
+++ b/src/main/java/jterm/command/Dir.java
@@ -77,7 +77,7 @@ public class Dir {
             isAbsoluteDirectory = true;
         } else if (JTerm.IS_WIN) {
             // Window paths are all relative unless they start with the drive string or a backslash
-            if (newDirectory.matches("((?i)(?s)[A-Z]):.*") || newDirectory.charAt(0) == '\\') {
+            if (newDirectory.matches("((?i)(?s)[A-Z]):.*") || newDirectory.charAt(0) == '\\' || newDirectory.charAt(0) == '/') {
                 isAbsoluteDirectory = true;
             } else {
                 isAbsoluteDirectory = false;

--- a/src/main/java/jterm/command/Dir.java
+++ b/src/main/java/jterm/command/Dir.java
@@ -71,40 +71,51 @@ public class Dir {
         File dir = new File(newDirectory);
         File newDir = new File(JTerm.currentDirectory + newDirectory);
 
+        // Perform checks to see if the path is relative to the current or is an absolute path
         boolean isAbsoluteDirectory;
-        if ( (JTerm.IS_UNIX && newDirectory.charAt(0) == '/') || (JTerm.IS_WIN && newDirectory.startsWith(".+:")) ) {
+        if (JTerm.IS_UNIX && newDirectory.charAt(0) == '/') {
             isAbsoluteDirectory = true;
+        } else if (JTerm.IS_WIN) {
+            // Window paths are all relative unless they start with the drive string or a backslash
+            if (newDirectory.matches("((?i)(?s)[A-Z]):.*") || newDirectory.charAt(0) == '\\') {
+                isAbsoluteDirectory = true;
+            } else {
+                isAbsoluteDirectory = false;
+            }
         } else {
             isAbsoluteDirectory = false;
         }
 
-        // Store each subdirectory into an array
         String subdirectories[];
-        if (isAbsoluteDirectory) {
-            if (JTerm.IS_WIN) {
-                subdirectories = newDirectory.split("\\\\");
-            } else {
-                subdirectories = newDirectory.split("/");
-            }
-        } else {
-            if (JTerm.IS_WIN) {
-                newDirectory = JTerm.currentDirectory + "\\" + newDirectory;
-                subdirectories = newDirectory.split("\\\\");
-            } else {
-                newDirectory = JTerm.currentDirectory + "/" + newDirectory;
-                subdirectories = newDirectory.split("/");
-            }
+        if (!isAbsoluteDirectory) {
+            newDirectory = JTerm.currentDirectory + "/" + newDirectory;
         }
 
+        // Store each subdirectory into an array by splitting them based on forward or backslashes
+        // In the case of Windows, the forward slashes are replaced by backslashes
+        if (JTerm.IS_WIN) {
+            newDirectory = newDirectory.replace("/", "\\");
+            subdirectories = newDirectory.split("\\\\");
+        } else {
+            subdirectories = newDirectory.split("/");
+        }
+
+        // Holds the root location (either something like C: or \)
+        String windowsRootLocation = newDirectory.charAt(0) == '\\' ? "\\" : subdirectories[0];
+
         // For each subdirectory in the array, we build the new directory
-        Deque directoriesDeque = new LinkedList<>();
+        Deque<String> directoriesDeque = new LinkedList<>();
         for (int i = 0; i < subdirectories.length; i++) {
             if (subdirectories[i].equals(".") || subdirectories[i].trim().equals("")) {
                 continue;
             } else if (subdirectories[i].equals("..")) {
+                // Check if the drive name is the only directory left and avoid erasing it
+                if (directoriesDeque.size() == 1 && directoriesDeque.peek().equals(windowsRootLocation)) {
+                    continue;
+                }
                 // If ".." is in the directory path, remove the last directory
                 // added to the deque to "move up" the directory tree
-                if(!directoriesDeque.isEmpty()) {
+                else if (!directoriesDeque.isEmpty()) {
                     directoriesDeque.removeLast();
                 }
             } else {
@@ -114,10 +125,13 @@ public class Dir {
 
         StringBuilder sb = new StringBuilder();
 
-        if (JTerm.IS_UNIX) {
+        if (JTerm.IS_WIN && newDirectory.charAt(0) == '\\') {
+            sb.append(windowsRootLocation);
+        } else if (JTerm.IS_UNIX) {
             sb.append("/");
         }
 
+        // Reconstruct the path string
         while (!directoriesDeque.isEmpty()) {
             sb.append(directoriesDeque.pop());
             if (JTerm.IS_WIN) {

--- a/src/main/java/jterm/command/Dir.java
+++ b/src/main/java/jterm/command/Dir.java
@@ -26,9 +26,7 @@ import org.apache.commons.io.FileUtils;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+import java.util.*;
 import java.util.function.Consumer;
 
 public class Dir {
@@ -73,28 +71,68 @@ public class Dir {
         File dir = new File(newDirectory);
         File newDir = new File(JTerm.currentDirectory + newDirectory);
 
-        if (newDirectory.equals("/")) {
-            newDirectory = "/";
-        } else if (newDirectory.equals(".")) {
-            return;
-        } else if (newDirectory.equals("..")) {
-            if (JTerm.currentDirectory.equals("/")) {
-                return;
-            } else {
-                //TODO: Fix this to actually remove a directory level
-                newDirectory = JTerm.currentDirectory.substring(0, JTerm.currentDirectory.length() - 2);
-                newDirectory = newDirectory.substring(0, newDirectory.lastIndexOf('/'));
-            }
-        } else if (newDir.exists() && newDir.isDirectory()) {
-            newDirectory = JTerm.currentDirectory + newDirectory;
-        } else if ((!dir.exists() || !dir.isDirectory()) && (!newDir.exists() || !newDir.isDirectory())) {
-            JTerm.out.println(TextColor.ERROR, "ERROR: Directory \"" + newDirectory + "\" either does not exist or is not a valid directory.");
-            
-            return;
+        boolean isAbsoluteDirectory;
+        if ( (JTerm.IS_UNIX && newDirectory.charAt(0) == '/') || (JTerm.IS_WIN && newDirectory.startsWith(".+:")) ) {
+            isAbsoluteDirectory = true;
+        } else {
+            isAbsoluteDirectory = false;
         }
 
-        if (!newDirectory.endsWith("/")) {
-            newDirectory += "/";
+        // Store each subdirectory into an array
+        String subdirectories[];
+        if (isAbsoluteDirectory) {
+            if (JTerm.IS_WIN) {
+                subdirectories = newDirectory.split("\\\\");
+            } else {
+                subdirectories = newDirectory.split("/");
+            }
+        } else {
+            if (JTerm.IS_WIN) {
+                newDirectory = JTerm.currentDirectory + "\\" + newDirectory;
+                subdirectories = newDirectory.split("\\\\");
+            } else {
+                newDirectory = JTerm.currentDirectory + "/" + newDirectory;
+                subdirectories = newDirectory.split("/");
+            }
+        }
+
+        // For each subdirectory in the array, we build the new directory
+        Deque directoriesDeque = new LinkedList<>();
+        for (int i = 0; i < subdirectories.length; i++) {
+            if (subdirectories[i].equals(".") || subdirectories[i].trim().equals("")) {
+                continue;
+            } else if (subdirectories[i].equals("..")) {
+                // If ".." is in the directory path, remove the last directory
+                // added to the deque to "move up" the directory tree
+                if(!directoriesDeque.isEmpty()) {
+                    directoriesDeque.removeLast();
+                }
+            } else {
+                directoriesDeque.add(subdirectories[i]);
+            }
+        }
+
+        StringBuilder sb = new StringBuilder();
+
+        if (JTerm.IS_UNIX) {
+            sb.append("/");
+        }
+
+        while (!directoriesDeque.isEmpty()) {
+            sb.append(directoriesDeque.pop());
+            if (JTerm.IS_WIN) {
+                sb.append("\\");
+            } else {
+                sb.append("/");
+            }
+        }
+
+        newDirectory = sb.toString();
+
+        if ((!dir.exists() || !dir.isDirectory()) && (!newDir.exists() || !newDir.isDirectory())) {
+            JTerm.out.println(TextColor.ERROR, "ERROR: Directory \"" + newDirectory + "\" either does not exist or is not a valid directory.");
+
+            return;
         }
 
         // It does exist, and it is a directory, so just change the global working directory variable to the input

--- a/src/main/java/jterm/gui/Terminal.java
+++ b/src/main/java/jterm/gui/Terminal.java
@@ -21,8 +21,8 @@ public class Terminal extends JFrame implements KeyListener {
 
     public Terminal() {
         TextColor.initGui();
-        JTerm.IS_WIN = false;
-        JTerm.IS_UNIX = true;
+        // JTerm.IS_WIN = false;
+        // JTerm.IS_UNIX = true;
 
         setContentPane(contentPane);
         // call onCancel() when cross is clicked

